compiler/noirc_frontend/src/elaborator/expressions.rs

```diff
@@ -542,30 +553,46 @@ impl<'context> Elaborator<'context> {
         let expr_id = self.interner.push_expr(expr);
         self.interner.push_expr_location(expr_id, span, self.file);
 
-        let typ = match self.infix_operand_type_rules(&lhs_type, &operator, &rhs_type, span) {
+        let result = self.infix_operand_type_rules(&lhs_type, &operator, &rhs_type, span);
+        let typ =
+            self.handle_operand_type_rules_result(result, &lhs_type, Some(trait_id), expr_id, span);
+
+        self.interner.push_expr_type(expr_id, typ.clone());
+        (expr_id, typ)
+    }
+
+    fn handle_operand_type_rules_result(
+        &mut self,
+        result: Result<(Type, bool), TypeCheckError>,
+        operand_type: &Type,
+        trait_id: Option<TraitMethodId>,
+        expr_id: ExprId,
+        span: Span,
+    ) -> Type {
+        match result {
             Ok((typ, use_impl)) => {
                 if use_impl {
+                    let trait_id =
+                        trait_id.expect("ice: expected some trait_id when use_impl is true");
```

Comment:
Because `trait_id` is `Option` in the case of an infix operator, if `use_impl` is true we assume `triat_id` won't be `None`... but ideally this `expect` shouldn't exist. Or is this fine? Or maybe there's another way to do this...

Reply:
I guess it is fine for now. I'd rather reduce repeated code here than break this back into two large code blocks we'd have to remember to change together just to get rid of an `expect`.